コメント機能と日報の実装

ShotaTsuru opened the pull request:
# What
コメント機能と日報の実装

# Why
レビューをいただくため

ShotaTsuru committed:
reportのscaffold

ShotaTsuru committed:
modified db/schema.rb

ShotaTsuru committed:
コメント機能マイグレート〜コントローラー作成中

ShotaTsuru committed:
コメントの投稿機能createアクション実装まで

ShotaTsuru committed:
コメント修正削除機能実装、部分テンプレートによるビューの共通化

ShotaTsuru committed:
追加機能のi18n日本語化、日時設定、削除時の依存関係設定

ShotaTsuru committed:
rubocopの修正

ShotaTsuru committed:
翻訳漏れの修正

JunichiIto commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+# rubocop:disable all
```
rubocopは勝手に無効化しちゃだめですー。ここはrubocopが指摘した問題点を解決してください。

どうしても解決できない場合はrubocopが通らない状態で提出して、「ここがどうしても解決できませんでした」とひとことコメントを入れて提出してください。

ShotaTsuru replied:
create_アクションのコードが長すぎますよっというエラーでこれ以上は短くできないなと勝手に判断して処理しておりました失礼いたしました！！

もし解決できない場合は、コメントを入れて提出いたします！🙇‍♂️🙇‍♂️

JunichiIto commented on app/views/books/show.html.erb:
```diff
@@ -20,5 +20,7 @@
   <%= image_tag(@book.picture_url) if @book.picture.present? %>
 </p>
 
+<%= render 'comments/new', locals: {article: @article, comment: @comment, comments: @comments} %>
```
localsなしで、直接article以下の引数を指定できるはずです。

ShotaTsuru replied:
できました！その辺りの仕様をしっかり理解しきっておりませんでした😭😭🙇‍♂️🙇‍♂️

JunichiIto commented on app/views/comments/_new.html.erb:
```diff
@@ -0,0 +1,35 @@
+              <%= comment.user.name %>
+            </div>
+            <div class="comment_date_time">
+              <%= comment.created_at%>
```
`%>`の手前にスペースを入れましょう。

ShotaTsuru replied:
前回同様のミスをしており注意不足でございました。。🙇‍♂️

JunichiIto commented on app/views/comments/_new.html.erb:
```diff
@@ -0,0 +1,35 @@
+<h2><%= Comment.model_name.human %></h2>
+<%= form_with model: @comment, url: polymorphic_url([@article, @comment]), local: true do |f| %>
```
`polymorphic_url`なしで`[@article, @comment]`だけでも動いたりしませんかね？

ShotaTsuru replied:
あぁ！！おっしゃる通りです！！polymorphic_urlmメソッドにばかり気を取られmodelオプションがインスタンスの中身からurlを自動生成してくれる事が頭からすっ飛んでいました！

JunichiIto commented on app/controllers/books_controller.rb:
```diff
@@ -11,7 +11,11 @@ def index
+  def show
+    @comment = Comment.new
+    @article = @book
+    @comments = @book.comments
```
複数レコードが返ってくる場合はorderを指定した方がいいです。

参考 [【Rails】開発環境とテストコード上（または本番環境）でデータの並び順が異なる場合の原因と対処方法 \- Qiita](https://qiita.com/jnchito/items/90c28c3f3e856add0a82)

ShotaTsuru replied:
複数レコードの取得に対しては習慣づけるようにいたします！！

JunichiIto commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,63 @@
+      @report = Report.find(params[:report_id])
+      @comment = @report.comments.build(comment_params)
+      @comments = @report.comments
+      respond_to do |format|
```
jsonのレスポンスは想定していないはずなので、

```ruby
if @comment.save
  redirect_to report_path(@report), notice: t('controllers.common.notice_create', name: Comment.model_name.human) }
else
```

のようにrespond_toは外してしまってください。（他の箇所も同様）

ShotaTsuru replied:
承知致しました！

scalffoldでできたものに対してわからないから下手にいじるよりそっとしておこうとしていました。ちゃんと自分で意図して使用できるよう深掘りして実装に必要か判断すべきでした！！

JunichiIto commented on app/views/comments/_new.html.erb:
```diff
@@ -0,0 +1,35 @@
+          </div>
+          <div class="comment_status_box">
+            <%= link_to t('views.common.edit'), polymorphic_url([:edit, @article, comment]) %>
+            <%= link_to t('views.common.destroy'), polymorphic_url([@article, comment]), method: :delete %>
```
コメントの編集と削除はコメント投稿者本人に限定した方がいいと思いますー。

ShotaTsuru replied:
承知致しました！

JunichiIto commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,79 @@
+
+  # Only allow a list of trusted parameters through.
+  def report_params
+    params.require(:report).permit(:title, :text).merge(user_id: current_user.id)
```
何かの間違いでupdate時にuserが変わると怖いので、ここでは何もせず、createアクションで、

```ruby
@report.user = current_user
```

のように書くことをオススメしますー。

ShotaTsuru replied:
保全性？完全性？を考えると確かに！！🤩🤩
今まで何も気にせずパラムスでマージするものだと思っておりました！

承知致しました！コメント機能も同様に修正いたします！

JunichiIto commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,79 @@
+  before_action :set_report, only: %i[show edit update destroy]
+  before_action :user_only_contributor, only: %i[edit update destroy]
+
+  # GET /reports or /reports.json
```
jsonは想定していないはずなので、このへんのコメントも全部削除していきましょう。
scaffoldはあくまで叩き台なので、不要なコードやコメントは適宜削除していく必要があります。

ShotaTsuru replied:
承知致しました！

JunichiIto commented on app/models/book.rb:
```diff
@@ -2,4 +2,6 @@
 class Book < ApplicationRecord
   mount_uploader :picture, PictureUploader
+
+  has_many :comments, as: :commentable, dependent: :destroy
```
各モデルに必須入力チェックのようなバリデーションを追加してみてください。
その上で、各モデルの入力画面から未入力状態でサブミットしたときにちゃんと意図通りに動くかテストしてみましょう。

ShotaTsuru replied:
承知致しました！

要件によるかもとは感じているのですが、各モデルの必須入力チェックバリデーションを追加した際には、db側でもエラーを通知するようマイグレーションファイルにもnull false制約をかけることがセットになるのが一般的でしょうか？（今回はまだ制約をかける記述は触っていません。）

JunichiIto replied:
> 要件によるかもとは感じているのですが、各モデルの必須入力チェックバリデーションを追加した際には、db側でもエラーを通知するようマイグレーションファイルにもnull false制約をかけることがセットになるのが一般的でしょうか？

good questionですね！

理想を言えば制約も追加すべし、なのですが、業務で開発するアプリケーションであれば、すでに既存のレコードがあると、「すでにあるレコードに対しては、どのタイミングでどの値を設定するのか？」という問題が出てきます。（何も考えずにマイグレーションを実行すると、既存のレコードがnot null制約に引っかかってエラーが起きる）

ただ、具体的にどうやるかは、ケースバイケースでかなりたくさんのバリエーションがあり、話し出すとキリがないのでここでは割愛します。
やむを得ずRails側のバリデーションだけ付けて、DBの制約は諦めるケースもあります。

今回もいったんDB制約は付けない、という形でOKです。

ShotaTsuru replied:
なるほど！把握致しました！

JunichiIto reviewed (commented):
コメントしましたー。

ShotaTsuru committed:
部分テンプレートlocalsの記述を削除。
viewファイルの空白改行整理。
comment_new.html.erb内のform_withのurlを修正。
複数レコードの取得全てにorderを追加。
各controller内のrespond_toの記述を削除jsonの呼び出し記述を削除。
コメントの編集削除機能の投稿者限定処理の追加（view、controller）。
ストロングパラメーター内のuser_idのマージ記述を削除~createメソッド内へuser_id登録の処理を追加。
各モデルバリデーションの処理を追加。
不要なファイルを削除。

ShotaTsuru committed:
rubocopの自動修正。
commentのcreateメソッド処理を切り分けすることで、createアクションのコードが長すぎるrubocopの指摘を回避。

JunichiIto commented on app/models/book.rb:
```diff
@@ -4,4 +4,10 @@ class Book < ApplicationRecord
+    validates :title
+    validates :memo
+    validates :author
+  end
```
`validates :title, :memo, :author, presence: true` みたいな書き方もいけるはずですー。

ShotaTsuru replied:
確かにそちらの方がスマートです！承知致しました！

JunichiIto commented on app/views/comments/_new.html.erb:
```diff
@@ -1,8 +1,22 @@
-<%= form_with model: @comment, url: polymorphic_url([@article, @comment]), local: true do |f| %>
+
+<%= form_with model: [@article, @comment], local: true do |f| %>
+  
```
この空行はなくてもいいかも

JunichiIto commented on app/views/reports/edit.html.erb:
```diff
@@ -2,5 +2,5 @@
 <%= render 'form', report: @report %>
 
-<%= link_to t('views.common.show'), @report %> |
+<%= link_to t('views.common.show'), @report %> 
```
細かいですが、`|`を消した理由って何かありますか？

ShotaTsuru replied:
`|`を消したことすらに気づいておりませんでした😅
修正いたします！

ShotaTsuru replied:
`|`を消したことすらに気づいておりませんでした😅
修正いたします！

JunichiIto commented on app/views/reports/show.html.erb:
```diff
@@ -13,7 +13,9 @@
 </p>
 
-<%= render 'comments/new', locals: {article: @article, comment: @comment, comments: @comments} %>
+<%= render 'comments/new', {article: @article, comment: @comment, comments: @comments} %>
```
Rubyは最後の引数がハッシュだった場合に`{}`が省略できますー。（チェリー本とかを読んでみて下さい）

ShotaTsuru replied:
すみません！単純な消し忘れでございました🙇‍♂️chengesみてちゃんと変更点チェックして提出いたします。。

JunichiIto commented on app/views/users/_users.html.erb:
```diff
@@ -16,7 +16,7 @@
         <th>
           <% if user.avatar.attached? %>
-            <%= image_tag user.avatar.variant(resize_to_fill: [30, 30])%>
+            <%= image_tag user.avatar.variant(resize_to_fill: [30, 30]) %>
```
これ、元のコードからですね。。。どうもすいません 😓

ShotaTsuru replied:
元コードだと意識せず触っておりました😅揚げ足取りのような変更になり申し訳ありません🙇‍♂️🙇‍♂️🙇‍♂️🙇‍♂️

JunichiIto replied:
いえいえ、こちらのミスなので気になさらずに！

JunichiIto commented on db/schema.rb:
```diff
@@ -35,7 +35,7 @@
 
   create_table "active_storage_variant_records", force: :cascade do |t|
-    t.integer "blob_id", null: false
+    t.bigint "blob_id", null: false
```
schema.rb、実行環境によっては新しいmigrationファイルがないのに勝手に変更されることがあるんですよね。。
今回はこのままでもいいですが、チーム開発するときは意図せずschema.rbが書き換えられるのを避けた方がいいです。（こういう問題に遭遇したら先輩プログラマに対処法を相談してください）

ShotaTsuru replied:
把握しておきます！

JunichiIto commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,74 @@
+    redirect_to polymorphic_path(@comment.commentable) unless @comment.user == current_user
+  end
+
+  def book_comment_create
```
メソッド名は動詞で始まる方がいいので、`create_book_comment`の方がベターです。

参考
[モデルやメソッドに名前を付けるときは英語の品詞に気をつけよう \- Qiita](https://qiita.com/jnchito/items/459d58ba652bf4763820)

JunichiIto commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,74 @@
+  end
+
+  def book_comment_create
+    @book = Book.find(params[:book_id])
```
viewで使わないオブジェクトはローカル変数にしてスコープを狭くしましょう。
不必要にインスタンス変数を定義していないかコード全体を再チェックしてみてください

ShotaTsuru replied:
承知致しました！

ShotaTsuru replied:
`@book`並びに`@report`は保存処理失敗時のrenderビューで使用するために、インスタンス変数として残しました！
destroyアクション時のインスタンス変数は使用しないので削除いたしました！
全体を見てそれ以外修正箇所が思いつかず見落としがあれば申し訳ありません。

JunichiIto replied:
変更内容はOKです。
ただし、`@book`と`@report`はバッククオートで囲まないとbookさんとreportさんにメンションを飛ばすことになる（というか、たぶん飛ばしてる）ので要注意です〜。

ShotaTsuru replied:
あぁ！ご迷惑おかけしてるやつですね！修正しました！今後気をつけます！

JunichiIto commented on app/controllers/books_controller.rb:
```diff
@@ -4,14 +4,16 @@ class BooksController < ApplicationController
-  def show; end
+  def show
+    @comment = Comment.new
+    @article = @book
```
`@comment = @book.comments.new`のようにすると、`@comment.commentable`で`@book`が取れるようになります。
これにより、viewで`@article`を使わないようにすることができそうです。

ShotaTsuru replied:
なるほどです！確かにコメントコントローラーでも同じ呼び出し方でとっていました！！

JunichiIto commented on app/views/comments/_new.html.erb:
```diff
@@ -0,0 +1,51 @@
+<h2><%= Comment.model_name.human %></h2>
+
+<%= form_with model: [@article, @comment], local: true do |f| %>
```
`[@comment.commentable, @comment]`のようにすれば、`@article`を一掃できそうな気がします。
（そもそも`@article`という名前も「Articleクラスがあるの？」と誤解を招きそうな名前なので、あまり良くないかもです）

ShotaTsuru replied:
承知致しました！

JunichiIto reviewed (commented):
コメントしましたー。

ShotaTsuru committed:
レビューの修正
・バリデーションの記述法を簡略化。
・_new.html.erb内の空白行の削除。
・ビューファイルの省略できるハッシュ修正。
・コメントコントローラー内のメソッド名を修正
・viewで使用しないインスタンス変数をローカル化
・インスタンス変数@articleを削除

ShotaTsuru committed:
使用していないファイルの削除

JunichiIto commented on app/views/comments/_new.html.erb:
```diff
@@ -0,0 +1,49 @@
+<% end %>
+
+<div class="comment_box">
+  <% if @comments.present? %>
```
`<% comments = comment.commentable.comments.order(:id) %>`のようにすれば`@comments`も不要になりそうですね。
ただ、viewにはなるべくロジックを持ち込むな、という意見もあるので、これがベストかどうかは何ともいいがたいです。

ShotaTsuru replied:
なるほどです！今回は`<% comments = comment.commentable.comments.order(:id) %>`こちらの記述もある事を記憶に残すため修正しておきます！

JunichiIto commented on app/views/comments/_new.html.erb:
```diff
@@ -0,0 +1,49 @@
+<h2><%= Comment.model_name.human %></h2>
+<%= form_with model: [@comment.commentable, @comment], local: true do |f| %>
```
パーシャルviewではインスタンス変数を直接参照せず、パラメータとして渡しましょう。

ShotaTsuru replied:
承知致しました！

JunichiIto commented on app/views/comments/_new.html.erb:
```diff
@@ -0,0 +1,49 @@
+          <% if comment.user == current_user %>
+            <div class="comment_status_box">
+              <%= link_to t('views.common.edit'), polymorphic_url([:edit, @comment.commentable, comment]) %>
+              <%= link_to t('views.common.destroy'), polymorphic_url([@comment.commentable, comment]), method: :delete %>
```
`polymorphic_url`がまだ残ってるようです。

ShotaTsuru replied:
modelオプションを使っているform_withだけ省略ができると勘違いしておりました！
link_toのurlも省略ができることがただいま把握致しました！ここら辺の省略の仕組みが曖昧なので少し深掘りしておきます！

JunichiIto reviewed (commented):
コメントしましたー。

ShotaTsuru committed:
レビューの修正
・controllerとview内の@commentsを削除。
・パーシャル内のインスタンス変数を修正。

JunichiIto commented on app/views/comments/_new.html.erb:
```diff
@@ -1,11 +1,11 @@
 <h2><%= Comment.model_name.human %></h2>
-<%= form_with model: [@comment.commentable, @comment], local: true do |f| %>
-  <% if @comment.errors.any? %>
+<%= form_with model: [comment.commentable, comment], local: true do |f| %>
```
`comment.commentable`がたくさん出てくるので、

```ruby
commentable = comment.commentable
```

のように変数に入れてDRYにした方がいいかも。

ShotaTsuru replied:
承知致しました部分テンプレートの受け渡しの際にcommentableにしておきます！

JunichiIto commented on app/views/comments/_new.html.erb:
```diff
@@ -19,27 +19,27 @@
 <div class="comment_box">
-  <% if @comments.present? %>
-    <% @comments.each do |comment| %>
+  <% if comment.commentable.comments.order(:id).present? %>
```
```ruby
<% if comments = comment.commentable.comments.order(:id).presence %>
```

みたいな高等テクニックを使ってDRYにすることもできます。

https://qiita.com/jnchito/items/dedb3b889ab226933ccf#%E7%A9%BA%E3%81%AA%E3%82%89%E5%88%A5%E3%81%AE%E5%80%A4%E3%82%92%E4%BB%A3%E5%85%A5%E3%81%AE%E4%BB%A3%E3%82%8F%E3%82%8A%E3%81%ABpresence%E3%82%92%E4%BD%BF%E3%81%86

ShotaTsuru replied:
```ruby
 <% if comments = commentable.comments.order(:id).presence %>
    <% comments.each do |c| %>
```
present?だとture falseが返り変数にしても使い回せないので、presenceを使用して使い回すことでDRYにするということでしょうか？

うまく意図を読み取れておらず申し訳ありません！！

JunichiIto replied:
そういうことです。その理解で合ってます！

JunichiIto reviewed (commented):
コメントしましたー。

ShotaTsuru committed:
部分テンプレートへ受けわたすcomment.commentableを変数commentにしてDRYにする。
条件分岐時に使用する定義を流用するためにpresenceを使用し条件分岐を行う。

JunichiIto commented on app/views/books/show.html.erb:
```diff
@@ -20,7 +20,7 @@
 </p>
 
-<%= render 'comments/new', comment: @comment %>
+<%= render 'comments/new', comment: @comment, commentable: @comment.commentable %>
```
うまく説明できてなくてすいません。僕が意図してたのはパラメータとして渡すことではなく、`_new.html.erb`の一番上の行などで

```erb
<% commentable = @comment.commentable %>
```

みたいにローカル変数を宣言することでした 🙏

ShotaTsuru replied:
固定概念のせいでviewの中にローカル変数を宣言するという発想がそもそもなかったです😂😂😂
なるほどです！ありがとうございます！！

JunichiIto commented on app/views/comments/_new.html.erb:
```diff
@@ -19,8 +19,8 @@
-  <% if comment.commentable.comments.order(:id).present? %>
-    <% comment.commentable.comments.order(:id).each do |c| %>
+  <% if comments = commentable.comments.order(:id).presence %>
+    <% comments.each do |c| %>
```
（前回コメントしたつもりだったのになぜか投稿できてなかった😓）

`c`だと短すぎるし、`comment`だと名前が被るし、悩ましいですね。
そもそも`_new.html.erb`が一覧を表示しようとしているのが問題のような気がします。パーシャルviewの責務を分割した方がいいですね。

`= render 'comments/comments', comments: comments` のように「コメント一覧表示専用のパーシャルview」を用意するとこの問題が解決できるかもしれません。

ShotaTsuru replied:
了解しました！

JunichiIto reviewed (commented):
コメントしました〜。

ShotaTsuru committed:
コメント一覧のパーシャルviewを追加
コメント一覧eachのブロック変数をcommentに変更

JunichiIto commented on app/views/comments/_new.html.erb:
```diff
@@ -20,29 +21,7 @@
+    <%= render 'comments/comments', comments: comments, commentable: commentable %>
   <% else %>
     <div><%= t("views.common.none_post", name: Comment.model_name.human) %></div>
   <% end %>
```
どこまで共通化すべきか、というのは要件によりますが、`comments/comments`内で条件分岐させて、「コメントがありません」のメッセージもパーシャル内で表示させる、という考え方もあります。いちおうご参考までに。

ShotaTsuru replied:
承知致しました！ありがとうございます！

JunichiIto reviewed (approved):
1点コメントしましたが、参考情報なのでコード自体はこれで問題ありません！合格です〜 🙆‍♂️
